### PR TITLE
Parallel assembly (partial)

### PR DESCRIPTION
**Pull request by babrodtk**
This pull request cuts the total runtime by around 20% compared to current master in my experiments, and produces the exact same results as master. It runs part of the assembly in parallel using OpenMP.

One caveat is that some printout may suffer from race conditions when using multiple threads, but that will hopefully be taken care of with a proper logger.

The pull request should not affect anyone, unless they enable USE_OPENMP in CMake. Then, the only side effect should be faster execution time, but more CPU use whilst running a simulation. 

I have not tested this in conjunction with MPI, but it should in theory be compatible.

### CODE REVIEW

**Comment by akva2** [comment]
is this really the proper granularity to use threads at? couldn't a grid coloring approach be taken instead? i suspect you are wasting a lot of time administering threads.

**Comment by babrodtk** [comment]
I agree with you that grid coloring would be preferable and "the right way" to do it, but that requires a lot more effort than what was put into this pull request. 

You are also correct  that there is a significant overhead to using OpenMP, but some of this is (at least somewhat) addressed by specifying the scheduling hint to OpenMP. 

This pull request still gives a significant (in my experiments) improvement on runtime with a minimal change to the code, and may therefore be of value until grid coloring or another approach is taken.

**Comment by akva2** [comment]
right. second question; can you avoid some by grouping stuff, i.e.

`
#pragma omp parallel
{
#pragma omp for
 for1
#pragma omp for
for 2
}
`

it's a bit hard to read from the diff..

**Comment by babrodtk** [comment]
I'm not quite sure what you mean by grouping stuff. If I haven't messed up, there should only be the addition of
```
#pragma omp parallel for schedule(X)
```
for independent (non-nested) for loops, which either loop over the Jacobi blocks, or the independent wells.

**Comment by alfbr** [comment]
I am a bit uncertain about this approach. Keep in mind that the MPI version of the code in coming, and will distribute the assembly part too. More to the performance part, did you test on Norne? How many threads do you need to get the benefit (Amdahl's law you know, we usually do not have many idle cpu's available on nodes used for reservoir simulation)?

**Comment by babrodtk** [comment]
I see your concern, and the upcoming MPI version may very well make this superfluous.

On my machine, with 6 ht cores, It appears that 4 threads gives the most bang for the buck. More threads give only smaller improvements, and fewer threads gives noticeably lower performance.

It was tested on norne, giving me the 20% decrease in runtime (about 3600 s for master, and around 3000 for this code.)

On 2. sep. 2015, at 15.34, Alf Birger Rustad <notifications@github.com<mailto:notifications@github.com>> wrote:


I am a bit uncertain about this approach. Keep in mind that the MPI version of the code in coming, and will distribute the assembly part too. More to the performance part, did you test on Norne? How many threads do you need to get the benefit (Amdahl's law you know, we usually do not have many idle cpu's available on nodes used for reservoir simulation)?

-

**Comment by akva2** [comment]
i see the formatting got a bit weird.

you can avoid some thread sleep/wake cycles by grouping openmp statements.

threads are awoken at a #pragma omp parallel statement
then you can have multiple parallel constructs within the parallel block. sometimes this can speed things up a bit.

**Comment by alfbr** [comment]
Cool! Given that this is very little intrusive in the code (so it should not bother developers much), I am all in favour.  Please consider limiting the maximum number of threads used to four.

**Comment by atgeirr** [comment]
> Please consider limiting the maximum number of threads used to four.

That is usually handled with environment variables, which is not that friendly I guess. We could add a "number_of_openmp_threads" parameter somewhere (where?) that is only asked for if we have openmp and defaults to 4, I think that would be both simple and flexible.

**Comment by alfbr** [comment]
Ah, I explained porly. I meant that default number of threads should be number of cores available, but no more than four. Allowing users to dictate more threads through the standard OMP_NUM_THREADS variable is fine of course. Most computers has four or more cores these days, so simply going with four as default is fine too by me.

**Comment by babrodtk** [comment]
By default, OpenMP creates one thread per virtual core (i.e., for a 6 core with ht, it typically creates 12 threads). By my experiments, 4 would be sufficient to get a decent speedup without spending too much CPU, so I implemented so that it defaults to max(4, #num_cores) *unless* OMP_NUM_THREADS has been set. 

Of course, adding a separate parameter might be just as good, so that the logic would read something like
```
if (!param.has("number_of_openmp_threads")) 
  //set default to max(4, #cores)
else 
  //set to param value
```

If this is deemed better than calling
```
OMP_NUM_THREADS=42 flow <deck-filename> 
```
I'd be happy to add the required logic

**Comment by alfbr** [comment]
OMP_NUM_THREADS is the right choice, custom variables should be avoided. Everything looks fine from my end.

**Comment by babrodtk** [comment]
> you can avoid some thread sleep/wake cycles by grouping openmp statements.

There were two cases where this could be done. There was no noticeable change in performance by grouping these statements and the diff became a bit more involved, so I suggest leaving it as is.

**Comment by blattms** [comment]
But if we use MPI (one process per core) and in addition OpenMP with
one thread per core that will bite us, right?
-- 
Join us at the DUNE User Meeting Sept. 28-29, 2015, Heidelberg, Germany
http://users.dune-project.org/projects/dune-user-meeting-2015/wiki

**Comment by alfbr** [comment]
Sure, so I do not envision using this multithreading in MPI runs. Still, it is a nice boost we all can get now in anticipation of an MPI version :-)

File: examples/flow.cpp [hunk 1]
```diff
@@ -148,6 +152,20 @@ try
         std::cout << "**********************************************************************\n\n";
     }
 
+#ifdef _OPENMP
+    if (!getenv("OPM_NUM_THREADS")) {
```
Comment (atgeirr) [inline examples/flow.cpp]: Typo: OPM -> OMP, right?

Comment (babrodtk) [inline examples/flow.cpp]: Uhm, well spotted... Thanks

**Comment by atgeirr** [comment]
Now that OpenMP is off by default here I will merge this PR, thanks!

File: examples/flow.cpp [hunk 2]
```diff
@@ -150,6 +154,20 @@ try
         std::cout << "**********************************************************************\n\n";
     }
 
+#ifdef _OPENMP
+    if (!getenv("OMP_NUM_THREADS")) {
+        //Default to max(4,#cores) threads,
+        //not number of cores (unless ENV(OMP_NUM_THREADS) is defined)
+        int num_cores = omp_get_num_procs();
+        int num_threads = std::min(4, num_cores);
```
Comment (bska) [inline examples/flow.cpp]: The comment says `std::max()`, but the code says `std::min()`.  Which is the intended semantics?

Comment (atgeirr) [inline examples/flow.cpp]: The intended semantics is max... Thanks for catching! Will fix.

Comment (babrodtk) [inline examples/flow.cpp]: Ok, a bit of a clumsy comment there (in the source code):

if you have 2 cores, you should use at most 2 threads, if you have 89 cores, you should use at most 4 threads. Code is right, comment is wrong :)

Comment (atgeirr) [inline examples/flow.cpp]: Hrrmmm... that is what I would like to say as well \<blush\>...

**Pull request merged**